added alert api

vaukalak opened the pull request:
**This patch solves the following problem**

When developing mobile application, there a lot of cases when `Alert` api is used. Anyway it's not available in `react-native-web` and developer have to rewrite a lot of code. This patch solves the issue.

**Test plan**

Added to storie book.

**This pull request**

- [x] includes documentation
- [ ] includes tests
- [x] includes an interactive example
- [ ] includes screenshots/videos

necolas commented:
Thank you! I'll get around to reviewing this soon

necolas commented on docs/apis/Alert.md:
```diff
@@ -0,0 +1,39 @@
+By default it'll look as ios alert, anyway there's possibility to style it. 
+## Methods
+
+static **setStyle**(config: Object)
```
Please remove the `setStyle` API and document `alert`

vaukalak replied:
@necolas why don't you like `setStyle` API? the app can run, for example, on ios/web platforms in this case alert similar to android native alert will not make a lot of sense. I think it could be a good idea to export STYLE_ANDROID and STYLE_IOS. And keep possibility to switch to desired style.. BTW, android alerts has different styles from version to version, could you please point me to the image of alert you'd like to have?

necolas commented on src/apis/Alert/index.js:
```diff
@@ -0,0 +1,153 @@
+
+let instance;
+
+let styles = {
```
styles should be registered with `StyleSheet.create`

necolas commented on src/apis/Alert/index.js:
```diff
@@ -0,0 +1,153 @@
+
+let styles = {
+  background: {
+    position: 'absolute',
```
You can spread `StyleSheet.absoluteFillObject` instead

necolas commented on src/apis/Alert/index.js:
```diff
@@ -0,0 +1,153 @@
+  render() {
+    const { visible, message, title, buttons } = this.state;
+    if (!visible) {
+      return <View />;
```
should be able to return `null` here

necolas commented on src/apis/Alert/index.js:
```diff
@@ -0,0 +1,153 @@
+            {
+              buttons.map(
+                ({ text, onPress }, i) => (
+                  <TouchableOpacity
```
In terms of accessibility, focus needs to move to the first button when the alert opens and return to the previously focused element after it closes

vaukalak replied:
@necolas Not sure, how to implement it..

vaukalak commented:
@necolas 
1 - what should we do with the `setStyle` api? I think this feature would be very use full in web. On mobile platforms RN displays native alert, which are intuitive for users, on web, it should be probably allowed, to show users the alert which follows general application style.
2 - not sure where to indicate api to close on background press: can't find anything related [in the doc](https://facebook.github.io/react-native/docs/alert.html)
3 - not sure what do you mean by focusing on the first button.
4 - need a screenshot of android alert you'd like to be as default style.
5 - after update test are successful on my mac, but fails on travis, can't understand the reason from the log.

thanks!

vaukalak commented:
@necolas sorry for disturbing, but could you please reply to my question?

necolas commented:
1 - remove `setStyle`
2 - https://github.com/facebook/react-native/blob/master/Libraries/Alert/Alert.js
3 - focus needs to move to the first button when the Alert opens
4 - you can see the default Android style in a RN app - rnplay.org
5 - https://travis-ci.org/necolas/react-native-web/builds/190528888

KjellConnelly commented:
@vaukalak How is this coming along? I'm very excited to have a matching Alert dialog :)

dsernst commented:
@vaukalak: This looks really great. Like @KjellConnelly said, it will be super useful to have a matching Alert api.

Another way to start using this sooner is to ship it as its own `node_module`. How would you feel about that?

vaukalak commented:
@dsernst I had a lot of work to do for latest months. so will fix soon. yep, removing `setStyle` Api will be sad. I think to create an issue for discussion on that, after this PR will be merged. will fix everything on this PR this week.

vaukalak commented:
Hey @necolas I've updated the PR, now build pass. Also removed `setStyle` api, and added possibility to close on background click. However do not understand completely what I have to do with p.3 (sorry, I'm noob in web :) ). Can we leave that point as improvement for the future, just to make that PR proceed faster?
Right now sounds that rnplay doesn't have web version. Different versions of android have different alert design. could you just point me to a screenshot of design you'd like to achieve? Than I'll implement it right away.

vaukalak commented:
@necolas I've updated alert style to be android like. please let me know if there's something else to do on that PR.

code-by commented:
Hi!
Could you please merge it, I need it urgent for my project :)

MoOx commented:
@code-by maybe you should instead consume branch directly instead of saying "I am in a hurry".

vadzim commented:
+1

vaukalak commented:
Hey @necolas completed also p. 3 (focus on first button). so all notes your made should be fixed now. Is there anything more to do on this PR?

MovingGifts commented:
Hey @necolas, is this going to get merged soon?

bschellekens commented:
Hi guys!
I too am developing a web version of our mobile react-native app, and love react-native-web.
However, we are using a lot of Alert.alert's, so this commit sounds perfect. 
Any progress in getting this branch merged?

JulianMayorga commented:
Hi everyone! ✋ 🎃 

Just came here because I tried to integrate react-router 4 native router into a react-native-web app, but it did not work because that router [depends on react-native Alert API](https://github.com/ReactTraining/react-router/blob/master/packages/react-router-native/NativeRouter.js#L4).

I want to help in this PR. @necolas is there anything I can do in order for you to accept it?

Being able to use react-router with react-native-web would be awesome. Technically it is possible to just use react-router's MemoryRouter, because NativeRouter is just a wrapper over it. But the documented way to use react-native with react-router is with react-router-native.

vaukalak commented:
Hey @necolas sorry, for disturbing, but this PR was here for more than 1 year. Is there anything I can do to make it progress? I do not think, that I'm the only one who has ports of native apps for the web. And absence of this Api, is painfull and increase time to port from native to the web... Thanks.

lijy91 commented:
Maybe you should consider call multiple problem.

```
Alert.alert('alert1');
Alert.alert('alert2');
Alert.alert('alert3');
```
As above, it only shows `alert3`.

johntran commented:
@vaukalak @necolas What are some ways I can help out with getting this in? :) I can definitely start with converting this to use React Portals, but are there anything else?

vaukalak commented:
@johntran I'd convert this to portals right now. But I'm also not sure, about all the list of things to happen to get this merged.

sfcgeorge commented:
I assume tests would be the final step.

ReactXP has this API so perhaps that could provide implementation inspiration https://microsoft.github.io/reactxp/docs/apis/alert.html
